Enhance filesystem dock tooltips

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/175

Added this to the tooltips:
- file size
- resource type (if resource)
- bigger preview (if texture or image)
![godot windows tools 64_BQPpwf9LeL](https://user-images.githubusercontent.com/2223172/180087999-28c26bfe-e1c6-40ca-a777-b5f2807e723d.gif)

KoBeWi committed:
Enhance filesystem dock tooltips

YeldhamDev commented:
The borders of the tooltips for images seem broken.

KoBeWi commented:
What do you mean?
The bigger tooltip in the GIF just went outside recording area 🙃

YeldhamDev commented:
> What do you mean?

The left border has a much smaller gap than the rest when hovering images.

KoBeWi commented:
Borders are managed by the default tooltip container, so it doesn't seem like my fault. I could maybe hack it by adding an empty control.

YeldhamDev commented:
It would be better if we found the root issue rather than implementing a hack. Still, not a stopper for this PR.

fire-forge commented:
What do you think about adding previews for more resources there, using the same system the FileSystem uses to get previews in grid mode? That way more resources, like scenes, scripts, gradients, materials, fonts, etc. could have tooltip previews.

![image](https://user-images.githubusercontent.com/67974470/180102160-108ea93e-99a8-46db-971b-103658d8a314.png)

KoBeWi commented:
I'd leave it for later. The PR lays a nice foundation for such additions.

fire-forge commented on editor/filesystem_dock.cpp:
```diff
@@ -2120,6 +2136,84 @@ void FileSystemDock::set_file_list_display_mode(FileListDisplayMode p_mode) {
+			if (image.is_valid()) {
+				preview = ImageTexture::create_from_image(image);
+
+				Label *label = memnew(Label(TTR("Dimensions:") + vformat(" %d x %d", preview->get_width(), preview->get_height())));
```
```suggestion
				Label *label = memnew(Label(TTR("Dimensions:") + vformat(String::utf8(" %s×%s"), preview->get_width(), preview->get_height())));
```

See https://github.com/godotengine/godot/pull/60008#discussion_r847653853

timothyqiu replied:
I would suggest moving the translation into the vformat call. So that the colon & space after it can be properly localized. e.g. CJK usually prefer removing the space after a full-width colon.

```suggestion
				Label *label = memnew(Label(vformat(TTR("Dimensions: %d×%d"), preview->get_width(), preview->get_height())));
```

Same to other hardcoded whitespace here.

fire-forge commented on editor/filesystem_dock.cpp:
```diff
@@ -2120,6 +2136,84 @@ void FileSystemDock::set_file_list_display_mode(FileListDisplayMode p_mode) {
+		if (ClassDB::is_parent_class(type, "Texture2D")) {
+			preview = ResourceLoader::load(p_path);
+			if (preview.is_valid()) {
+				Label *label = memnew(Label(TTR("Dimensions:") + vformat(" %d x %d", preview->get_width(), preview->get_height())));
```
```suggestion
				Label *label = memnew(Label(TTR("Dimensions:") + vformat(String::utf8(" %s×%s"), preview->get_width(), preview->get_height())));
```

groud commented:
I don't am not sure the preview implementation is acceptable as loading a resource can be costly. In that case it's only images, but I think I would have preferred that this PR uses the built in resource preview system.

It should even simplify the implementation, as it's fairly easy to use. You can check for uses of `EditorResourcePreview::get_singleton()->queue_resource_preview` to check how it works.

KoBeWi commented:
I tried this
```C++
TextureRect *tr = memnew(TextureRect);
EditorResourcePreview::get_singleton()->queue_resource_preview(p_path, tr, "set_texture", Variant());
```
but it doesn't have effect 🤔

groud commented:
Yeah I agree with the fact that `EditorResourcePreview` could also generate some metadata to be displayed along with the picture or something like that.
If you don't want to implement that right now and maybe to avoid freezes, maybe you could load the texture in the `_tooltip_thumbnail_done` callback? With some luck the texture will still be kept loaded and it would avoid the loading freeze.

KoBeWi commented:
It doesn't work. When the preview is cached it doesn't even load the texture. I can look into implementing preview metadata, but I don't think it really blocks this PR.

reduz commented:
All welcome the new Godot Operating System overlord.

reduz commented on editor/filesystem_dock.cpp:
```diff
@@ -2120,6 +2147,63 @@ void FileSystemDock::set_file_list_display_mode(FileListDisplayMode p_mode) {
+Control *FileSystemDock::create_tooltip_for_path(const String &p_path) const {
+	if (DirAccess::exists(p_path)) {
+		// No tooltip for directory.
+		return nullptr;
```
Well, this is kind of hacky and you know how this story continues in Godot. After the PR is merged, you get a proposal opened:

"Nice resource file info tooltip you have there, can I do this for my custom resource type from my extension?" 

So I think this should be a bit more pluggable. So maybe it would be good to make a tiny class: `EditorResourceFileTooltipPlugin` and register them directly into the FileSystem dock? (via EditorPlugin API).

KoBeWi replied:
That's too much work, I'd rather wait for someone to actually request it (the code won't be difficult to replace).

mhilbrunner commented:
Editor PR review meeting: feature is wanted and approved, @KoBeWi wants to improve texture loading first seperately.

KoBeWi commented:
Rebased after #64628. For now this PR still uses the hard-coded approach. I'll look into implementing EditorResourceFileTooltipPlugin.

KoBeWi commented:
Ok I added an API for creating resource tooltips and added a plugin for textures. We can add more plugins later (I have some interesting idea for audio files).

akien-mga commented:
Thanks!
